lldb/include/lldb/Core/DILAST.h

```diff
@@ -0,0 +1,459 @@
+/// very end, define all the types of AST nodes used by the DIL parser and
+/// expression evaluator. The DIL parser parses the input string and creates the
+/// AST parse tree from the AST nodes. The resulting AST node tree gets passed
+/// to the DIL expression evaluator, which evaluates the DIL AST nodes and
+/// creates/returns a ValueObjectSP containing the result.
+
+/// Base class for AST nodes used by the Data Inspection Language (DIL) parser.
+/// All of the specialized types of AST nodes inherit from this (virtual) base
+/// class.
+class DILASTNode {
+public:
+  DILASTNode(clang::SourceLocation location, NodeKind kind)
+      : m_location(location), m_kind(kind) {}
+  virtual ~DILASTNode() = default;
+
+  virtual void Accept(Visitor *v) const = 0;
+
+  virtual bool is_rvalue() const = 0;
+  virtual bool is_bitfield() const { return false; };
+  virtual bool is_context_var() const { return false; };
+  virtual bool is_literal_zero() const { return false; }
+  virtual uint32_t bitfield_size() const { return 0; }
+  virtual CompilerType result_type() const = 0;
+
+  clang::SourceLocation GetLocation() const { return m_location; }
+  NodeKind getKind() const { return m_kind; }
+
+  // The expression result type, but dereferenced in case it's a reference. This
+  // is for convenience, since for the purposes of the semantic analysis only
+  // the dereferenced type matters.
+  CompilerType GetDereferencedResultType() const;
+
+private:
+  clang::SourceLocation m_location;
+  const NodeKind m_kind;
+};
+
+using DILASTNodeUP = std::unique_ptr<DILASTNode>;
+
+class ErrorNode : public DILASTNode {
+public:
+  ErrorNode(CompilerType empty_type)
+      : DILASTNode(clang::SourceLocation(), NodeKind::eErrorNode),
+        m_empty_type(empty_type) {}
+  void Accept(Visitor *v) const override;
+  bool is_rvalue() const override { return false; }
+  CompilerType result_type() const override { return m_empty_type; }
+  CompilerType result_type_real() const { return m_empty_type; }
+
+  static bool classof(const DILASTNode *node) {
+    return node->getKind() == NodeKind::eErrorNode;
+  }
+
+private:
+  CompilerType m_empty_type;
+};
+
+class ScalarLiteralNode : public DILASTNode {
+public:
+  ScalarLiteralNode(clang::SourceLocation location, CompilerType type,
+                    Scalar value, bool is_literal_zero)
+      : DILASTNode(location, NodeKind::eScalarLiteralNode), m_type(type),
+        m_value(value), m_is_literal_zero(is_literal_zero) {}
+
+  void Accept(Visitor *v) const override;
+  bool is_rvalue() const override { return true; }
+  bool is_literal_zero() const override { return m_is_literal_zero; }
+  CompilerType result_type() const override { return m_type; }
+
+  auto value() const { return m_value; }
+
+  static bool classof(const DILASTNode *node) {
+    return node->getKind() == NodeKind::eScalarLiteralNode;
+  }
+
+private:
+  CompilerType m_type;
+  Scalar m_value;
+  bool m_is_literal_zero;
```

Comment:
I don't think so? I don't think having a special node type for a literal zero would simplify very much, and it would mean more code duplication (for handling the new node type). But I could be wrong...

Reply:
Hmm... actually, lets take a step back. Is it ever possible to have a `ScalarLiteralNode`, whose value is zero, but it is not a **literal** zero?

Reply:
I think so: If this were representing a Boolean false.

Reply:
Okay, so maybe the implementation of "literal zero" can be `m_scalar.IsZero() && !m_type.IsBoolean()`